### PR TITLE
[CLDC-20] Submit for other orgs

### PR DESCRIPTION
**Pull request by ghost**
https://digital.dclg.gov.uk/jira/browse/CLDC-20

Adds/updates 3 questions:
- submit for other orgs (only available to support users)
- managing agent
-- only shown if more than 2 available to pick from
-- if 1 value set automatically to that one
-- if 0 skipped
- housing provider
-- only shown if more than 2 available to pick from
-- if 1 value set automatically to that one
-- if 0 skipped

All work was put behind a feature flag. 

Screenshot 2022-11-14 at 14 54 22
Screenshot 2022-11-14 at 14 54 28
Screenshot 2022-11-14 at 14 54 39
Screenshot 2022-11-14 at 14 55 21

### CODE REVIEW

File: app/models/form/lettings/pages/housing_provider.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  def questions
+    @questions ||= [
+      Form::Lettings::Questions::HousingProvider.new(nil, nil, self),
+    ]
+  end
+
+  # For an organisation owns and manages ONLY their own housing stock AND NOT uses agents to manage the properties AND NOT manages housing stock of other organisations?
+  # In "set up this lettings log" no extra question should be shown
```
Comment (natdeanlewissoftwire) [inline app/models/form/lettings/pages/housing_provider.rb]: Took me a while to understand what this comment meant, maybe something like this is clearer:

#No extra question should be shown in "set up this lettings log" if:
#the organisation owns and manages ONLY their own housing stock
#AND doesn't have housing providers

File: app/models/form/lettings/pages/managing_organisation.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  def questions
+    @questions ||= [
+      Form::Lettings::Questions::ManagingOrganisation.new(nil, nil, self),
+    ]
+  end
+
+  # For an organisation owns and manages ONLY their own housing stock AND NOT uses agents to manage the properties AND NOT manages housing stock of other organisations?
+  # In "set up this lettings log" no extra question should be shown
```
Comment (natdeanlewissoftwire) [inline app/models/form/lettings/pages/managing_organisation.rb]: Similar to above, but this is different, right?

#No extra question should be shown in "set up this lettings log" if:
#the organisation owns and manages ONLY their own housing stock
#AND doesn't have managing agents

File: config/initializers/feature_toggle.rb [hunk 3]
```diff
@@ -20,4 +20,10 @@ def self.location_toggle_enabled?
 
     false
   end
+
+  def self.managing_for_other_user_enabled?
+    return true unless Rails.env.production?
+
+    false
+  end
```
Comment (natdeanlewissoftwire) [inline config/initializers/feature_toggle.rb]: I've updated the rest of these toggles, hope you agree this is simpler:

```suggestion
  def self.managing_for_other_user_enabled?
    !Rails.env.production?
  end
```

File: app/models/form/lettings/questions/housing_provider.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    true
+  end
+
+  def hidden_in_check_answers?(_log, user = nil)
+    @current_user = user
+
+    return false unless @current_user
+    return false if @current_user.support?
+
+    # Hide when less than 2 housing providers
```
Comment (natdeanlewissoftwire) [inline app/models/form/lettings/questions/housing_provider.rb]: small thing - is this comment necessary? I think it's pretty self explanatory

File: app/models/form/lettings/questions/managing_organisation.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    true
+  end
+
+  def hidden_in_check_answers?(_log, user = nil)
+    @current_user = user
+
+    return false unless @current_user
+    return false if @current_user.support?
+
+    # Hide when less than 2 managing_agents
```
Comment (natdeanlewissoftwire) [inline app/models/form/lettings/questions/managing_organisation.rb]: same as above

File: db/seeds.rb [hunk 6]
```diff
@@ -18,8 +18,28 @@
...
+    address_line1: "2 Marsham Street",
+    address_line2: "London",
+    postcode: "SW1P 4DF",
+    holds_own_stock: true,
+    other_stock_owners: "None",
+    managing_agents_label: "None",
+    provider_type: "LA",
+  )
+  managing_agent2 = Organisation.find_or_create_by!(
+    name: "Managing Agent 1",
```
Comment (natdeanlewissoftwire) [inline db/seeds.rb]: Managing Agent 2*

File: app/models/form/lettings/pages/housing_provider.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    return false unless current_user
+    return true if current_user.support?
+    return true unless current_user.organisation.holds_own_stock?
+
+    housing_providers = current_user.organisation.housing_providers
+
+    return false if housing_providers.count.zero?
+    return true if housing_providers.count > 1
+
+    log.update!(owning_organisation: housing_providers.first)
```
Comment (kosiakkatrina) [inline app/models/form/lettings/pages/housing_provider.rb:26]: What if the org holds it's own stock and has 1 housing provider? Could it not be an owning organisation of a log?

File: app/models/form/lettings/questions/housing_provider.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    true
+  end
+
+  def housing_providers
+    @housing_providers ||= if current_user.support?
+                             Organisation.all
+                           else
+                             current_user.organisation.housing_providers
+                           end
+  end
```
Comment (jamesrose) [inline app/models/form/lettings/questions/housing_provider.rb]: I _think_ you can build your options for select and improve the performance of it.

```suggestion
  def housing_providers_answer_options
    @housing_provider_answer_options ||= if current_user.support?
                             Organisation
                           else
                             current_user.organisation.housing_providers
                           end.pluck(:id, :name).to_h
  end
```

File: app/models/form/lettings/questions/housing_provider.rb [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+  def hidden_in_check_answers?(_log, user = nil)
+    @current_user = user
+
+    return false unless @current_user
+    return false if @current_user.support?
+
+    # Hide when less than 2 housing providers
+    housing_providers.count < 2
```
Comment (jamesrose) [inline app/models/form/lettings/questions/housing_provider.rb]:
```suggestion
    housing_providers_answer_options.count < 2
```

File: app/models/form/lettings/questions/housing_provider.rb [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    return answer_opts unless current_user
+
+    if !current_user.support? && current_user.organisation.holds_own_stock?
+      answer_opts[current_user.organisation.id] = "#{current_user.organisation.name} (Your organisation)"
+    end
+
+    housing_providers.select(:id, :name).each_with_object(answer_opts) do |organisation, hsh|
+      hsh[organisation.id] = organisation.name
+      hsh
+    end
```
Comment (jamesrose) [inline app/models/form/lettings/questions/housing_provider.rb]:
```suggestion
    answer_opts.merge(housing_providers_answer_options)
```

**Review by jamesrose — COMMENTED** [review]
Up to you if you want to address my comments. I think there's an argument of clarity for keeping it, but you get a bit of a performance and brevity bump with my suggestion. Your call.

File: app/models/form/lettings/pages/housing_provider.rb (hunk 7 above)
Comment (ghost) [reply app/models/form/lettings/pages/housing_provider.rb:26]: if the org holds its own stock and has one housing provider, we are the log `owning_organisation` to that housing provider

**Review by jamesrose — APPROVED** [review]
Sweet. Nice work!

File: app/models/form/lettings/pages/housing_provider.rb (hunk 7 above)
Comment (natdeanlewissoftwire) [reply app/models/form/lettings/pages/housing_provider.rb:26]: Couldn't the org be its own `owning_organisation` for this log in that case? It owns housing stock after all so must want to submit logs for these properties sometimes right?

**Pull request merged**